src/ralph/virtual/tests/test_openstack.py

```diff
@@ -0,0 +1,198 @@
+        )
+        project1 = CloudProject.objects.create(
+            name='project1',
+            cloudprovider=self.cloud_provider,
+            project_id='project_id_1',
+        )
+        CloudProject.objects.create(
+            name='project2',
+            cloudprovider=self.cloud_provider,
+            project_id='project_id_2',
+        )
+        CloudProject.objects.create(
+            name='project3',
+            cloudprovider=self.cloud_provider,
+            project_id='project_os_id1',
+        )
+
+        host = CloudHost.objects.create(
+            host_id='host_id1',
+            hostname='host1',
+            parent=project1,
+            cloudflavor=self.cloud_flavor
+        )
+        host.ipaddress_set.create(address='2.2.3.4')
+
+        IPAddress.objects.create(address='1.2.3.4')
+
+        self.cmd = Command()
+        self.cmd._get_cloud_provider()
+        self.cmd.openstack_projects = OPENSTACK_DATA
+        self.cmd.openstack_flavors = OPENSTACK_FLAVOR
+        self.cmd._get_ralph_data()
+
+    def test_check_get_ralph_data(self):
+        ralph = self.cmd.ralph_projects
+        self.assertEqual(ralph['project_id_1']['name'], 'project1')
+        self.assertEqual(ralph['project_id_1']['cloudprovider'],
+                         self.cloud_provider)
+
+    def test_check_add_ip(self):
+        """Check if IP is added correctly to a host"""
+        host = CloudHost.objects.get(host_id="host_id1")
+        self.cmd._add_ip(host, '9.9.9.9', False)
+        ips = map(lambda x: x.address, host.ipaddress_set.all())
+        self.assertIn('9.9.9.9', list(ips))
+
+    def test_check_add_ip_assign(self):
+        """Check if IP is assigned correctly to a host"""
+        host = CloudHost.objects.get(host_id="host_id1")
+        self.cmd._add_ip(host, '1.2.3.4', False)
+        ips = map(lambda x: x.address, host.ipaddress_set.all())
+        self.assertIn('1.2.3.4', list(ips))
+
+    def test_check_add_servers(self):
+        """Check if servers are added and modified correctly"""
+        project = CloudProject.objects.get(project_id='project_id_1')
+        self.cmd._add_servers(TEST_HOSTS, project)
+        for host_id in TEST_HOSTS.keys():
+            host = CloudHost.objects.get(host_id=host_id)
+            ips = map(lambda x: x.address, host.ipaddress_set.all())
+            self.assertEqual(host.hostname, TEST_HOSTS[host_id]['hostname'])
+            self.assertEqual(
+                datetime.strptime(
+                    TEST_HOSTS[host_id]['created'],
+                    self.cmd.DATETIME_FORMAT
+                ),
+                host.created,
+            )
+            self.assertIn(TEST_HOSTS[host_id]['tag'], host.tags.names())
+            for ip in TEST_HOSTS[host_id]['ips']:
+                self.assertIn(ip, list(ips))
+
+    def test_check_add_flavor(self):
+        """Check if flavors are added and modified correctly"""
+        for flavor_id in OPENSTACK_FLAVOR:
```

Comment:
use ``.items()`` for ``OPENSTACK_FLAVOR`` then you can iter key and value at the same time
``for flavor_id, flavor in OPENSTACK_FLAVOR.items()``